features/environment.py

```diff
@@ -1,294 +1,340 @@
-                self.handle.kill()
-            time.sleep(1)
-            if not kill and time.time() - start_time > timeout:
-                kill = True
-        self.handle = None
-        if self.log_file and not self.log_file.closed:
-            self.log_file.close()
-        if self.work_directory:
-            shutil.rmtree(self.work_directory)
-            self.work_directory = None
+        self.stop(timeout=timeout)
+        if self._work_directory:
+            shutil.rmtree(self._work_directory)
 
-    @staticmethod
-    def cleanup_service_tree():
+    @abc.abstractmethod
+    def query(self, key):
+        """ query for a value of a given key """
+
+    @abc.abstractmethod
+    def cleanup_service_tree(self):
         """ clean all contents stored in the tree used for the tests """
-        r = None
+
+
+class EtcdController(AbstractDcsController):
+
+    """ handles all etcd related tasks, used for the tests setup and cleanup """
+
+    def __init__(self, output_dir):
+        super(EtcdController, self).__init__('etcd', tempfile.mkdtemp(), output_dir)
+        self._client = etcd.Client()
+
+    def _start(self):
+        return subprocess.Popen(["etcd", "--debug", "--data-dir", self._work_directory],
+                                stdout=self._log, stderr=subprocess.STDOUT)
+
+    def query(self, key):
+        try:
+            return self._client.get('/{0}/{1}'.format(self._CLUSTER_NODE, key)).value
+        except etcd.EtcdKeyNotFound:
+            return None
+
+    def cleanup_service_tree(self):
         try:
-            r = requests.delete(EtcdController.ETCD_CLEANUP_URL)
-            if r and not r.ok:
-                assert False,\
-                    "request to cleanup the etcd contents was not successfull: status code {0}".format(r.status_code)
-        except requests.exceptions.RequestException as e:
+            self._client.delete('/' + self._CLUSTER_NODE, recursive=True)
+        except (etcd.EtcdKeyNotFound, etcd.EtcdConnectionFailed):
+            return
+        except Exception as e:
             assert False, "exception when cleaning up etcd contents: {0}".format(e)
 
-    @staticmethod
-    def _is_running():
+    def _is_running(self):
         # if etcd is running, but we didn't start it
         try:
-            r = requests.get(EtcdController.ETCD_VERSION_URL)
-            running = (r and r.ok and b'etcdserver' in r.content)
-        except requests.ConnectionError:
-            running = False
-        return running
+            return bool(self._client.machines)
+        except Exception:
+            return False
+
+
+class ZooKeeperController(AbstractDcsController):
+
+    """ handles all zookeeper related tasks, used for the tests setup and cleanup """
+
+    def __init__(self, output_dir):
+        super(ZooKeeperController, self).__init__('zookeeper', None, output_dir)
+        self._client = kazoo.client.KazooClient()
+
+    def _start(self):
+        pass  # TODO: implement later
```

Comment:
zookeeper is started up as a "service" from .travis.yml
The same approach as for etcd or patroni doesn't work with zookeeper, because `start script` exists immediately.